Layout iteration - Leading story promo 1008px+ layout

Bopchy opened the pull request:
Resolves #2208 

**Overall change:** _Implement layout iteration designs for the leading story promo for screen widths of 1008px+._

**Code changes:**

- _Create psammead-leading-story-promo component._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Automated jest tests added (for new features) or updated (for existing features)
- [x] This PR requires manual testing

Commit:
Create story-promo-feature package.

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Get grid to work.

Commit:
Remove unused code. Update CHANGELOG, README.

Commit:
Add secondary story promo. Remove comments. Add prop validation.

Commit:
Add snapshot test.

Commit:
Rename package.

Commit:
Rename package. Update test & snapshot. Tag component as alpha.

Commit:
Merge branch '2208-leading-story-promo-1008px+-layout' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Add leading story promo + secondary promo example.

Commit:
Remove mediaIndicator stuff for now.

Commit:
Update README.

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout

Bopchy commented on packages/components/psammead-leading-story-promo/src/index.stories.jsx:
```diff
@@ -0,0 +1,109 @@
+      );
+    },
+  });
+/* eslint-enable react/prop-types */
```
Will make this section less repetitive

Bopchy replied:
Made this less repetitive by adding `generateInfo()`

DenisHdz commented on packages/components/psammead-leading-story-promo/src/index.jsx:
```diff
@@ -0,0 +1,21 @@
+import Grid from '@bbc/psammead-grid';
+
+const LeadingStoryPromo = ({ image, info }) => (
+  <Grid columns={{ group4: 8 }} enableGelGutters>
```
I think we should not be using the `psammead-grid` within the component. This will be added in Simorgh as a wrapper, when we set up the layout that will be using this component.

This component should be isolated and styled as we do with the current Top/regular story promos, with `display: grid` and fallbacks.

Bopchy replied:
Ok

DenisHdz commented on packages/components/psammead-leading-story-promo/src/index.stories.jsx:
```diff
@@ -0,0 +1,109 @@
+      const Img = buildImg();
+
+      return (
+        <Grid columns={{ group4: 8 }} enableGelGutters>
```
We should also have a default story, with only the component, without any Grid layout.

Commit:
Create LeadingPromoWrapper, ImageGridItem, TextGridItem. Use them in a story (generateFallbackStory).

Commit:
Create uses-grid.jsx. Update stories. Update snapshot.
- Move work that used psammead-grid to create component layout to `uses-grid.jsx`.
- Update both stories to use the `display: grid` implementation.

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Merge branch '2208-leading-story-promo-1008px+-layout' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

DenisHdz commented on packages/components/psammead-leading-story-promo/CHANGELOG.md:
```diff
@@ -0,0 +1,5 @@
+
+| Version | Description |
+|---------|-------------|
+| 1.0.0-alpha.0| [PR#2399](https://github.com/bbc/psammead/pull/2399) Initial creation of package. |
```
From this https://github.com/bbc/psammead/pull/2428, we will be setting the first version to `0.1.0-alpha.1`.

DenisHdz commented on packages/components/psammead-leading-story-promo/src/index.stories.jsx:
```diff
@@ -0,0 +1,109 @@
+  });
+/* eslint-enable react/prop-types */
+
+storiesOf('Components|LeadingStoryPromo', module)
```
Since this is another type of promo, I would place it under the current `StoryPromo` folder.
`storiesOf('Components|StoryPromo/LeadingStoryPromo', module)`

DenisHdz commented on packages/components/psammead-leading-story-promo/src/index.jsx:
```diff
@@ -0,0 +1,80 @@
+    display: block;
+    width: initial;
+
+    @media (min-width: ${GEL_GROUP_4_SCREEN_WIDTH_MIN}) {
```
Do we know if this component will be used just above 1008px? If so, I would remove this media query to keep the grid styles by default.

Otherwise, we might need to tweak it differently under 1008px, rather than just using `display:block`.

Bopchy replied:
Mentioned on Slack - ```The styling will be applied only at 1008+ as below that breakpoint we will be maintaining the one column list approach such that all leading promos within a section will look like a top story promo.```

DenisHdz commented on packages/components/psammead-leading-story-promo/src/index.jsx:
```diff
@@ -0,0 +1,80 @@
+const TextGridItem = styled.div`
+  display: inline-block;
+  vertical-align: top;
+
```
We will also need fallbacks as we do with the `StoryPromo`.

DenisHdz commented on packages/components/psammead-leading-story-promo/src/uses-grid.jsx:
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
```
Where is this file used?

Bopchy replied:
This is not being used. Removed it.

DenisHdz commented on packages/components/psammead-leading-story-promo/src/index.stories.jsx:
```diff
@@ -0,0 +1,109 @@
+
+      return (
+        <Grid columns={{ group4: 8 }} enableGelGutters>
+          <Grid item columns={{ group4: 6 }}>
```
You should add the number of columns, for all the groups.

Example for your current layout:
```
<Grid
          columns={{
            group0: 8,
            group1: 8,
            group2: 8,
            group3: 8,
            group4: 8,
            group5: 8,
          }}
          enableGelGutters
        >
          <Grid
            item
            columns={{
              group0: 8,
              group1: 8,
              group2: 8,
              group3: 8,
              group4: 6,
              group5: 6,
            }}
          >
            <LeadingStoryPromo image={Img} info={Info} />
          </Grid>
          <Grid
            columns={{
              group0: 8,
              group1: 8,
              group2: 8,
              group3: 8,
              group4: 2,
              group5: 2,
            }}
          >
            <Grid
              item
              columns={{
                group0: 2,
                group1: 2,
                group2: 2,
                group3: 2,
                group4: 2,
                group5: 2,
              }}
            >
              {Img}
            </Grid>
            <Grid
              item
              columns={{
                group0: 6,
                group1: 6,
                group2: 6,
                group3: 6,
                group4: 2,
                group5: 2,
              }}
            >
              {Info}
            </Grid>
          </Grid>
        </Grid>
```

Commit:
Make suggested changes. Delete uses-grid.jsx.

Commit:
Add fallbacks for image and text grids.

Commit:
Update snapshot.

Commit:
Update README.

Bopchy commented:
The `leading promo + secondary promo` story doesn't look as expected on IE11 (that is, the secondary promo doesn't stay within the specified grid and instead maintains a full width), but looks ok on Chrome & FF. I think this is because of how `psammead-grid` currently works on IE

Commit:
Use withServicesKnob(). Create generateInfo().

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

j-pendlebury commented on packages/components/psammead-leading-story-promo/README.md:
```diff
@@ -0,0 +1,61 @@
+
+## Description
+
+The `LeadingStoryPromo` component is designed to be used on 'index' pages, which are pages that link to other articles/stories. It supports having info on the left with an image on the right. This can be a collection of any nodes.
```
I mean this is technically correct for most services, except for RTL

Bopchy replied:
Added a sentence on about LTR and RTL services in the usage section

j-pendlebury commented on packages/components/psammead-leading-story-promo/src/index.test.jsx:
```diff
@@ -0,0 +1,26 @@
+    <Headline script={latin} service="news">
+      <Link href="https://www.bbc.co.uk/news">The headline of the promo</Link>
+    </Headline>
+    <Summary script={latin} service="news">
```
should we also have a test for a RTL service to make sure that we don't introduce any breaking changes in the future?

Bopchy replied:
Added a snapshot test, but it mostly just ensures the right typography is used. There is no structural change in page elements. I don't think it's necessary. Thoughts?

j-pendlebury replied:
my thinking was more that in the future when we make adjustments to the component without checking the RTL layout, at least the snapshot test _should_ flag up inconsistencies

Commit:
Update README. Add test. Update snapshot.`

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

j-pendlebury reviewed (approved):
LGTM 👍

sadickisaac commented on packages/components/psammead-leading-story-promo/README.md:
```diff
@@ -0,0 +1,61 @@
+
+## Description
+
+The `LeadingStoryPromo` component is designed to be used on 'index' pages, which are pages that link to other articles/stories. It displays both an image and info, where `info` can be a collection of any nodes.
```
Improve the definition of `Index` pages. See https://github.com/bbc/simorgh/pull/4299#discussion_r337458113

sadickisaac commented on packages/components/psammead-leading-story-promo/src/index.jsx:
```diff
@@ -0,0 +1,89 @@
+// (2 / 6) * 100 = 0.3333333333 = 33.33%
+
+const LeadingPromoWrapper = styled.div`
+  position: relative;
```
Why do we need to set the position to relative?

Bopchy replied:
This is not necessary. Removed it

AlistairGempf replied:
I think this `relative` prevents the link's `:before` from taking up the whole screen as can be seen here: 
![image](https://user-images.githubusercontent.com/43134742/67371641-98cb9100-f574-11e9-8880-b58ade54aad8.png)
I don't know how it acts when integrated into Simorgh though.

sadickisaac commented on packages/components/psammead-leading-story-promo/src/index.jsx:
```diff
@@ -0,0 +1,89 @@
+    display: grid;
+    grid-template-columns: repeat(6, 1fr);
+    grid-column-gap: ${GEL_SPACING_DBL};
+    grid-column-end: span 6;
```
Having `grid-column-end` here makes this difficult to use in other context where the desired span may not be 6.

Bopchy replied:
Removed `grid-column-end` definition

sadickisaac commented on packages/components/psammead-leading-story-promo/src/index.jsx:
```diff
@@ -0,0 +1,89 @@
+
+const ImageGridItem = styled.div`
+  display: inline-block;
+  vertical-align: top;
```
Is this necessary?

DenisHdz replied:
We need this to align the image to the top when using `display:inline-block;` on IE.

Bopchy replied:
No, it's not needed. Removed it.

DenisHdz commented on packages/components/psammead-leading-story-promo/CHANGELOG.md:
```diff
@@ -0,0 +1,5 @@
+
+| Version | Description |
+|---------|-------------|
+| 1.0.0-alpha.1 | [PR#2399](https://github.com/bbc/psammead/pull/2399) Initial creation of package. |
```
```suggestion
| 0.1.0-alpha.1 | [PR#2399](https://github.com/bbc/psammead/pull/2399) Initial creation of package. |
```

DenisHdz replied:
```suggestion
| 0.1.0-alpha.1 | [PR#2399](https://github.com/bbc/psammead/pull/2399) Initial creation of package. |
```

DenisHdz commented on packages/components/psammead-leading-story-promo/package-lock.json:
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "@bbc/psammead-leading-story-promo",
+  "version": "1.0.0-alpha.1",
```
```suggestion
  "version": "0.1.0-alpha.1",
```

DenisHdz commented on packages/components/psammead-leading-story-promo/package.json:
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "@bbc/psammead-leading-story-promo",
+  "version": "1.0.0-alpha.1",
```
```suggestion
  "version": "0.1.0-alpha.1",
```

DenisHdz commented on packages/components/psammead-leading-story-promo/src/index.jsx:
```diff
@@ -0,0 +1,89 @@
+
+const ImageGridColumnsFallback = css`
+  width: ${fourOfSixColumnsMaxWidthScaleable};
+  display: inline-block;
```
We've this `display: inline-block;` already below in the `ImageGridItem`.

Bopchy replied:
Removed this `display: inline-block;` and maintained the definition in `ImageGridItem`

DenisHdz commented on packages/components/psammead-leading-story-promo/src/index.jsx:
```diff
@@ -0,0 +1,89 @@
+const ImageGridItem = styled.div`
+  display: inline-block;
+  vertical-align: top;
+  position: relative;
```
Same with this line. I think we need it in the regular `Story Promo` because we have the media indicator displayed as an absolute, but not here.

Commit:
Update packages/components/psammead-leading-story-promo/CHANGELOG.md

Co-Authored-By: Denis Hernandez <46446236+DenisHdz@users.noreply.github.com>

Commit:
Update README, package-lock.json, package.json. Remove unnecessary styling.

Commit:
Update snapshots.

Commit:
Merge branch '2208-leading-story-promo-1008px+-layout' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Remove unused styling.

Commit:
Update snapshot.

Commit:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout

DenisHdz reviewed (approved):
We can move on with this component and apply any changes required later on, when we integrate it in Simorgh using the new layout 👍

DenisHdz committed:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout

Commit:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Merge branch '2208-leading-story-promo-1008px+-layout' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Fix font size.

DenisHdz committed:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout

Commit:
Merge branch 'latest' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Use StoryPromo in story.

Commit:
Merge branch '2208-leading-story-promo-1008px+-layout' of github.com:bbc/psammead into 2208-leading-story-promo-1008px+-layout

Commit:
Update test.

Commit:
Use <Headline /> for `leading promo + secondary story promo` story.

Commit:
Add isLeadingPromo = true

DenisHdz committed:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout

jamesbrumpton committed:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout

jamesbrumpton committed:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout

jamesbrumpton commented:
lgtm

jamesbrumpton committed:
Merge branch 'latest' into 2208-leading-story-promo-1008px+-layout
